File: latest/workshop1_m5/W01_M05_demo.py
"""
============================================================
Workshop 1 — Module 5: Typography, Layout & Composition
Python Demonstration Script
Data Visualization for Data Scientists — UNYT Tirana
============================================================
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
import os

os.makedirs("output", exist_ok=True)
np.random.seed(42)

# ── Global font settings (Tufte-friendly) ──────────────────
plt.rcParams.update({
    "font.family": "sans-serif",
    "font.size": 11,
    "axes.titlesize": 14,
    "axes.titleweight": "bold",
    "axes.labelsize": 11,
    "xtick.labelsize": 9,
    "ytick.labelsize": 9,
    "figure.dpi": 150,
})


# ── 1. Typographic Hierarchy ──────────────────────────────
products = ["Product F", "Product E", "Product D",
            "Product C", "Product B", "Product A"]
revenue = sorted([450, 380, 520, 290, 610, 340])

fig, ax = plt.subplots(figsize=(7, 5))
ax.barh(products, revenue, color="#1565C0", height=0.5, edgecolor="none")
for i, v in enumerate(revenue):
    ax.text(v + 8, i, f"${v}K", va="center", fontsize=8, fontweight="bold", color="#333")

# Five-level hierarchy
ax.set_title("Product Revenue: Q4 2024", fontsize=16, fontweight="bold", pad=8)
fig.text(0.13, 0.92, "Annual revenue in thousands USD, sorted descending",
         fontsize=11, color="grey", ha="left")
ax.set_xlabel("Revenue ($K)", fontsize=10, labelpad=8)
fig.text(0.13, 0.02, "Source: Internal sales database | Prepared: Jan 2025",
         fontsize=8, fontstyle="italic", color="#888", ha="left")

ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
ax.spines["left"].set_visible(False)
ax.tick_params(left=False, labelsize=9)
ax.set_xlim(0, 720)
plt.subplots_adjust(top=0.88, bottom=0.12, left=0.18, right=0.95)
plt.savefig("output/typographic_hierarchy_py.pdf", bbox_inches="tight")
plt.savefig("output/typographic_hierarchy_py.png", dpi=300, bbox_inches="tight")
plt.close()


# ── 2. Direct Labelling vs Legend ─────────────────────────
np.random.seed(7)
months = np.arange(1, 13)
series = {
    "Alpha": np.cumsum(np.random.normal(3, 4, 12)) + 40,
    "Beta":  np.cumsum(np.random.normal(2, 3, 12)) + 30,
    "Gamma": np.cumsum(np.random.normal(1, 5, 12)) + 20,
}
colors = {"Alpha": "#1565C0", "Beta": "#E53935", "Gamma": "#2E7D32"}

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(11, 4))

# Version A: legend
for name, data in series.items():
    ax1.plot(months, data, color=colors[name], linewidth=2, label=name)
ax1.legend(fontsize=8, frameon=True, loc="upper left")
ax1.set_title("Version A: Legend", fontsize=10, fontweight="bold")
ax1.set_xlabel("Month"); ax1.set_ylabel("Revenue ($K)")
ax1.spines["top"].set_visible(False); ax1.spines["right"].set_visible(False)

# Version B: direct labels
for name, data in series.items():
    ax2.plot(months, data, color=colors[name], linewidth=2)
    ax2.text(12.3, data[-1], name, fontsize=8, fontweight="bold",
             color=colors[name], va="center")
ax2.set_xlim(1, 15)
ax2.set_title("Version B: Direct Labels", fontsize=10, fontweight="bold")
ax2.set_xlabel("Month"); ax2.set_ylabel("Revenue ($K)")
ax2.spines["top"].set_visible(False); ax2.spines["right"].set_visible(False)

fig.suptitle("Direct Labelling Eliminates the Legend Shuttle",
             fontsize=12, fontweight="bold")
plt.tight_layout()
plt.savefig("output/direct_vs_legend_py.png", dpi=300, bbox_inches="tight")
plt.close()


# ── 3. Multi-Panel with GridSpec ──────────────────────────
np.random.seed(55)
fig = plt.figure(figsize=(10, 6))
gs = GridSpec(2, 3, figure=fig, hspace=0.4, wspace=0.35, height_ratios=[1, 1.3])

# (a) Bar chart
ax_a = fig.add_subplot(gs[0, 0])
cats = ["A", "B", "C", "D"]
vals = [42, 58, 35, 67]
ax_a.barh(cats, vals, color="#1565C0", height=0.5)
ax_a.set_title("(a) Revenue", fontsize=9, fontweight="bold")
ax_a.spines["top"].set_visible(False); ax_a.spines["right"].set_visible(False)
ax_a.tick_params(labelsize=7)

# (b) Scatter
ax_b = fig.add_subplot(gs[0, 1])
ax_b.scatter(np.random.uniform(0, 10, 40), np.random.uniform(0, 10, 40),
             s=20, c="#E53935", alpha=0.6)
ax_b.set_title("(b) Scatter", fontsize=9, fontweight="bold")
ax_b.spines["top"].set_visible(False); ax_b.spines["right"].set_visible(False)
ax_b.tick_params(labelsize=7)

# (c) Histogram
ax_c = fig.add_subplot(gs[0, 2])
ax_c.hist(np.random.normal(50, 15, 500), bins=25, color="#2E7D32",
          edgecolor="white", linewidth=0.3)
ax_c.set_title("(c) Distribution", fontsize=9, fontweight="bold")
ax_c.spines["top"].set_visible(False); ax_c.spines["right"].set_visible(False)
ax_c.tick_params(labelsize=7)

# (d) Full-width time series
ax_d = fig.add_subplot(gs[1, :])
t = np.arange(1, 25)
trend = np.cumsum(np.random.normal(2, 5, 24)) + 100
ax_d.plot(t, trend, color="#1565C0", linewidth=2)
ax_d.fill_between(t, trend, alpha=0.06, color="#1565C0")
ax_d.set_title("(d) 24-Month Trend (full width)", fontsize=9, fontweight="bold")
ax_d.set_xlabel("Month", fontsize=8)
ax_d.spines["top"].set_visible(False); ax_d.spines["right"].set_visible(False)
ax_d.tick_params(labelsize=7)

fig.suptitle("Multi-Panel Report: GridSpec Layout", fontsize=12, fontweight="bold")
plt.savefig("output/multipanel_report_py.pdf", bbox_inches="tight")
plt.savefig("output/multipanel_report_py.png", dpi=300, bbox_inches="tight")
plt.close()


# ── 4. Annotation Patterns ───────────────────────────────
np.random.seed(99)
months24 = np.arange(1, 25)
rev = np.cumsum(np.random.normal(2, 6, 24)) + 100
rev[11] = rev[10] - 25
rev[12:] = rev[12:] + 15

fig, ax = plt.subplots(figsize=(8, 4.5))

# Context band
ax.axvspan(14, 18, color="#E8F5E9", alpha=0.6)
ax.text(16, max(rev) + 5, "Recovery\nperiod", ha="center",
        fontsize=8, fontweight="bold", color="#2E7D32")

# Reference line
mean_rev = np.mean(rev)
ax.axhline(y=mean_rev, color="grey", linewidth=1, linestyle="--")
ax.text(24.5, mean_rev, f"Mean: {mean_rev:.0f}", fontsize=7, va="center", color="grey")

# Data
ax.plot(months24, rev, color="#1565C0", linewidth=2)

# Callout
ax.annotate("Product recall\n(Month 12)",
            xy=(12, rev[11]), xytext=(16, rev[11] - 18),
            fontsize=8, fontweight="bold", color="#C62828",
            arrowprops=dict(arrowstyle="->", color="#C62828", linewidth=1.5),
            bbox=dict(boxstyle="round,pad=0.3", facecolor="#FFEBEE",
                      edgecolor="#C62828", alpha=0.8))

ax.set_xlabel("Month", fontsize=9)
ax.set_ylabel("Revenue ($K)", fontsize=9)
ax.set_title("Three Annotation Patterns: Callout + Band + Reference",
             fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout()
plt.savefig("output/annotation_patterns_py.png", dpi=300, bbox_inches="tight")
plt.close()


print("── All M05 plots saved to output/ ──")

File: latest/workshop1_m5/W01_M05_demo.R
# ============================================================
# Workshop 1 — Module 5: Typography, Layout & Composition
# R Demonstration Script
# Data Visualization for Data Scientists — UNYT Tirana
# ============================================================

library(tidyverse)
# install.packages(c("patchwork", "ggrepel", "showtext"))
library(patchwork)
library(ggrepel)
dir.create("output", showWarnings = FALSE)


# ── 1. Typographic Hierarchy ──────────────────────────────
set.seed(42)
df <- tibble(
  product = paste("Product", LETTERS[1:6]),
  revenue = c(450, 380, 520, 290, 610, 340)
) |>
  mutate(product = fct_reorder(product, revenue))

p_typo <- ggplot(df, aes(x = product, y = revenue)) +
  geom_col(fill = "#1565C0", width = 0.6) +
  geom_text(aes(label = paste0("$", revenue, "K")),
            hjust = -0.15, size = 3, fontface = "bold") +
  coord_flip() +
  theme_minimal(base_size = 11) +
  theme(
    # Title hierarchy
    plot.title = element_text(size = 16, face = "bold",
                              margin = margin(b = 6)),
    plot.subtitle = element_text(size = 11, color = "grey50",
                                  margin = margin(b = 12)),
    plot.caption = element_text(size = 8, face = "italic", hjust = 0,
                                 margin = margin(t = 10)),
    # Axis text
    axis.title = element_text(size = 10),
    axis.text = element_text(size = 9),
    # Margins
    plot.margin = margin(t = 15, r = 30, b = 10, l = 10),
    panel.grid.major.y = element_blank()
  ) +
  labs(
    title = "Product Revenue: Q4 2024",
    subtitle = "Annual revenue in thousands USD, sorted descending",
    caption = "Source: Internal sales database | Prepared: Jan 2025",
    x = NULL, y = "Revenue ($K)"
  )

print(p_typo)
ggsave("output/typographic_hierarchy.png", p_typo, width = 7, height = 5, dpi = 300)
ggsave("output/typographic_hierarchy.pdf", p_typo, width = 7, height = 5)


# ── 2. Direct Labelling vs Legend ─────────────────────────
set.seed(7)
months <- 1:12
df_lines <- tibble(
  month = rep(months, 3),
  product = rep(c("Alpha", "Beta", "Gamma"), each = 12),
  revenue = c(
    cumsum(rnorm(12, 3, 4)) + 40,
    cumsum(rnorm(12, 2, 3)) + 30,
    cumsum(rnorm(12, 1, 5)) + 20
  )
)
colors <- c(Alpha = "#1565C0", Beta = "#E53935", Gamma = "#2E7D32")

# Version A: with legend
p_legend <- ggplot(df_lines, aes(x = month, y = revenue, color = product)) +
  geom_line(linewidth = 1.2) +
  scale_color_manual(values = colors) +
  theme_minimal(base_size = 10) +
  labs(title = "Version A: Legend", x = "Month", y = "Revenue ($K)", color = "Product")

# Version B: direct labels
endpoints <- df_lines |> filter(month == max(month))

p_direct <- ggplot(df_lines, aes(x = month, y = revenue, color = product)) +
  geom_line(linewidth = 1.2) +
  geom_text_repel(
    data = endpoints,
    aes(label = product),
    nudge_x = 0.8, size = 3.5, fontface = "bold",
    direction = "y", segment.color = "grey70"
  ) +
  scale_color_manual(values = colors) +
  scale_x_continuous(limits = c(1, 14)) +
  theme_minimal(base_size = 10) +
  theme(legend.position = "none") +
  labs(title = "Version B: Direct Labels", x = "Month", y = "Revenue ($K)")

p_labelling <- p_legend + p_direct +
  plot_annotation(title = "Direct Labelling Eliminates the Legend Shuttle")

print(p_labelling)
ggsave("output/direct_vs_legend.png", p_labelling, width = 11, height = 4, dpi = 300)


# ── 3. Multi-Panel Composition with patchwork ─────────────
set.seed(55)

# Panel (a): bar
p_a <- ggplot(df, aes(x = product, y = revenue)) +
  geom_col(fill = "#1565C0", width = 0.5) +
  coord_flip() +
  theme_minimal(base_size = 8) +
  theme(panel.grid.major.y = element_blank()) +
  labs(title = "(a) Revenue by Product", x = NULL, y = "$K")

# Panel (b): scatter
p_b <- ggplot(tibble(x = rnorm(80), y = rnorm(80)),
              aes(x = x, y = y)) +
  geom_point(color = "#E53935", alpha = 0.5, size = 1.5) +
  theme_minimal(base_size = 8) +
  labs(title = "(b) Scatter")

# Panel (c): histogram
p_c <- ggplot(tibble(val = rnorm(500, 50, 15)), aes(x = val)) +
  geom_histogram(fill = "#2E7D32", bins = 25, color = "white", linewidth = 0.3) +
  theme_minimal(base_size = 8) +
  labs(title = "(c) Distribution", x = "Value", y = "Count")

# Panel (d): full-width time series
p_d <- ggplot(tibble(
  month = 1:24,
  value = cumsum(rnorm(24, 2, 5)) + 100
), aes(x = month, y = value)) +
  geom_line(color = "#1565C0", linewidth = 1) +
  geom_area(fill = "#1565C0", alpha = 0.06) +
  theme_minimal(base_size = 8) +
  labs(title = "(d) 24-Month Trend", x = "Month", y = "Value")

# Compose
p_multi <- (p_a | p_b | p_c) / p_d +
  plot_annotation(
    title = "Multi-Panel Report: patchwork Layout",
    subtitle = "Top row: three metrics | Bottom: temporal context"
  ) +
  plot_layout(heights = c(1, 1.3))

print(p_multi)
ggsave("output/multipanel_report.png", p_multi, width = 10, height = 6, dpi = 300)
ggsave("output/multipanel_report.pdf", p_multi, width = 10, height = 6)


# ── 4. Annotation Patterns ───────────────────────────────
set.seed(99)
df_anno <- tibble(
  month = 1:24,
  revenue = cumsum(rnorm(24, 2, 6)) + 100
)
df_anno$revenue[12] <- df_anno$revenue[11] - 25
df_anno$revenue[13:24] <- df_anno$revenue[13:24] + 15

p_anno <- ggplot(df_anno, aes(x = month, y = revenue)) +
  # Context band (enclosure)
  annotate("rect", xmin = 14, xmax = 18, ymin = -Inf, ymax = Inf,
           fill = "#E8F5E9", alpha = 0.6) +
  annotate("text", x = 16, y = max(df_anno$revenue) + 5,
           label = "Recovery period", size = 3, fontface = "bold", color = "#2E7D32") +
  # Reference line
  geom_hline(yintercept = mean(df_anno$revenue), linetype = "dashed", color = "grey60") +
  annotate("text", x = 24.5, y = mean(df_anno$revenue),
           label = paste0("Mean: ", round(mean(df_anno$revenue))),
           size = 2.5, color = "grey50", hjust = 0) +
  # Data line
  geom_line(color = "#1565C0", linewidth = 1) +
  # Callout
  annotate("text", x = 15, y = df_anno$revenue[12] - 12,
           label = "Product recall\n(Month 12)", size = 2.8,
           fontface = "bold", color = "#C62828") +
  annotate("segment", x = 14, xend = 12.2,
           y = df_anno$revenue[12] - 8, yend = df_anno$revenue[12] + 1,
           arrow = arrow(length = unit(0.15, "cm")), color = "#C62828") +
  theme_minimal(base_size = 10) +
  labs(title = "Three Annotation Patterns",
       subtitle = "Callout + Context Band + Reference Line",
       x = "Month", y = "Revenue ($K)")

print(p_anno)
ggsave("output/annotation_patterns.png", p_anno, width = 8, height = 5, dpi = 300)

cat("\n── All M05 plots saved to output/ ──\n")
